src/main/java/org/spongepowered/api/event/cause/entity/damage/source/DamageSource.java

```diff
@@ -49,7 +49,7 @@
      *
      * @return A new builder
      */
-    static Builder builder() {
+    static <B extends Builder<DamageSource, B>> B builder() {
```

Comment:
What's the point of the generic if it's just making a `Builder` that would otherwise build a `DamageSource`? Is it magically going to return a builder that would build an `EntityDamageSource` or say a plugin's customized damage source builder?

Reply:
The point of this generic is that the `Builder` is not a leaf in interface inheritance; other `Builder` can extend this builder with a different bound specified for `A` (the returned built type). Thus, we just need to have one builder declared for each damage source declared in inheritance instead of two,

Reply:
Another reason to dump the generic self types.

Reply:
So @ST-DDT, if we dump generic self types, two ways to go:
1. Make all builder methods return void and use things like 
```java
Sponge.configure(DamageSource.builder(), $ -> {
    $.setMagic(true);
    $.setByPassesArmor(true);
    // More
});
```
instead.

2. Manually override the return type in each subclass (which is almost impossible).

Gabi and ST, thoughts?

Reply:
1) No.
2) Yes. With the help of an annotation processor that should be possible. It's a one time efford as the root builders dont change. It also has the added benefit that you can see all options of a builder by looking at its source.

Reply:
Sure, we could do that, just like what we did to sponge event factory. now, i think it is up to gabizou to decide.

Reply:
Annotation processor sounds fine, making it part of the build process to ensure we don't distribute jars with invalid builder types.

Reply:
So do add another gradle task that use spoon or one that uses asm to patch the built api jar? I don't know if spoon can modify existing generated source.

Reply:
I'd rather we use spoon over patching the built api jar.